src/MSBuildLocator/DotNetSdkLocationHelper.cs

```diff
@@ -78,124 +67,251 @@ public static IEnumerable<VisualStudioInstance> GetInstances(string workingDirec
+            if (Directory.Exists(hostFxrRoot))
             {
-                string fullPathToDotnetFromRoot = Path.Combine(dotnet_root, exeName);
-                if (File.Exists(fullPathToDotnetFromRoot))
+                // Agreed to load hostfxr from the highest version
+                var hostFxrAssemblyDirectory = Directory.GetDirectories(hostFxrRoot)
+                    .OrderByDescending(d => d)
+                    .FirstOrDefault();
+
+                if (hostFxrAssemblyDirectory != null)
                 {
-                    return realpath(fullPathToDotnetFromRoot) ?? fullPathToDotnetFromRoot;
+                    var hostfxrAssembly = Directory.GetFiles(hostFxrAssemblyDirectory)
+                        .Where(filePath => hostFxrRegex.IsMatch(Path.GetFileName(filePath)))
+                        .FirstOrDefault();
+
+                    if (hostfxrAssembly != null)
+                        return NativeLibrary.TryLoad(hostfxrAssembly, out var handle) ? handle : IntPtr.Zero;
                 }
             }
 
-            return null;
+            return IntPtr.Zero;
         }
 
-        private static IEnumerable<string> GetDotNetBasePaths(string workingDirectory)
+        private static string SdkResolutionExceptionMessage(string methodName) => $"Failed to find all versions of .NET Core MSBuild. Call to {methodName}. There may be more details in stderr.";
+        
+        /// <summary>
+        /// Determines the directory location of the SDK accounting for
+        /// global.json and multi-level lookup policy.
+        /// </summary>
+        private static string? GetSdkFromGlobalSettings(string workingDirectory)
         {
-            string dotnetPath = null;
-            bool isWindows = RuntimeInformation.IsOSPlatform(OSPlatform.Windows);
-            string exeName = isWindows ? "dotnet.exe" : "dotnet";
-
-            // First check for the DOTNET_ROOT environment variable, as it's often there as with, for example, dotnet format.
-            if (IntPtr.Size == 4)
+            string? resolvedSdk = null;
+            int rc = NativeMethods.hostfxr_resolve_sdk2(exe_dir: DotnetPath, working_dir: workingDirectory, flags: 0, result: (key, value) =>
             {
-                // 32-bit architecture
-                dotnetPath ??= FindDotnetFromEnvironmentVariable("DOTNET_ROOT(x86)", exeName);
-            }
-            else if (IntPtr.Size == 8)
+                if (key == NativeMethods.hostfxr_resolve_sdk2_result_key_t.resolved_sdk_dir)
+                {
+                    resolvedSdk = value;
+                }
+            });
+
+            if (rc != 0)
+                throw new InvalidOperationException(SdkResolutionExceptionMessage(nameof(NativeMethods.hostfxr_resolve_sdk2)));
+
+            return resolvedSdk;
+        }
+
+        private static string ResolveDotnetPath()
+        {
+            string? dotnetPath = GetDotnetPathFromROOT();
+
+            if (string.IsNullOrEmpty(dotnetPath))
```

Comment:
Most of these helper methods return null on failure, correct? Should be able to have a series of:

dotnetPath ??= thing();

-type calls. Should be easier to understand.

Reply:
I applied your recommendation in the else statement - looks great.

Here I would leave branching as it is because we have different probing logic for MsBuild.Location run from dotnet.exe or apphost.